### PR TITLE
[SPARK-50599][SQL] Create the DataEncoder trait that allows for Avro and UnsafeRow encoding

### PR DESCRIPTION
**Pull request by ericm-db**
### What changes were proposed in this pull request?

Currently, we use the internal byte representation to store state for stateful streaming operators in the StateStore. This PR introduces Avro serialization and deserialization capabilities in the RocksDBStateEncoder so that we can instead use Avro encoding to store state. This is currently enabled for the TransformWithState operator via SQLConf to support all functionality supported by TWS

### Why are the changes needed?

UnsafeRow is an inherently unstable format that makes no guarantees of being backwards-compatible. Therefore, if the format changes between Spark releases, this could cause StateStore corruptions. Avro is more stable, and inherently enables schema evolution.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Amended and added to unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 1]
```diff
@@ -76,17 +76,16 @@ private[sql] class RocksDBStateStoreProvider
...
+      val dataEncoder = getDataEncoder(
+        stateStoreEncoding, dataEncoderCacheKey, keyStateEncoderSpec, valueSchema)
 
       keyValueEncoderMap.putIfAbsent(colFamilyName,
-        (RocksDBStateEncoder.getKeyEncoder(keyStateEncoderSpec, useColumnFamilies,
-          Some(newColFamilyId), avroEnc), RocksDBStateEncoder.getValueEncoder(valueSchema,
-          useMultipleValuesPerKey, avroEnc)))
+        (RocksDBStateEncoder.getKeyEncoder(dataEncoder, keyStateEncoderSpec, useColumnFamilies,
+          Some(newColFamilyId)), RocksDBStateEncoder.getValueEncoder(dataEncoder, valueSchema,
+          useMultipleValuesPerKey)))
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:88]: nit: can you make these one parameter per line please? maybe define them outside of the putIfAbsent first

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 2]
```diff
@@ -387,17 +386,16 @@ private[sql] class RocksDBStateStoreProvider
...
+    val dataEncoder = getDataEncoder(
+      stateStoreEncoding, dataEncoderCacheKey, keyStateEncoderSpec, valueSchema)
 
     keyValueEncoderMap.putIfAbsent(StateStore.DEFAULT_COL_FAMILY_NAME,
-      (RocksDBStateEncoder.getKeyEncoder(keyStateEncoderSpec,
-        useColumnFamilies, defaultColFamilyId, avroEnc),
-        RocksDBStateEncoder.getValueEncoder(valueSchema, useMultipleValuesPerKey, avroEnc)))
+      (RocksDBStateEncoder.getKeyEncoder(dataEncoder, keyStateEncoderSpec,
+        useColumnFamilies, defaultColFamilyId),
+        RocksDBStateEncoder.getValueEncoder(dataEncoder, valueSchema, useMultipleValuesPerKey)))
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:398]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 3]
```diff
@@ -664,12 +689,29 @@ object RocksDBStateStoreProvider {
       avroOptions.stableIdPrefixForUnionType, avroOptions.recursiveFieldMaxDepth)
   }
 
+  /**
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:692]: IMHO all of this logic should live inside the AvroDataEncoder

**Review by brkyvz — CHANGES_REQUESTED** [review]
Overall looking good! Left some comments, we're missing some clean up that can be done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala [hunk 4]
```diff
@@ -60,15 +60,92 @@ sealed trait RocksDBValueStateEncoder {
  * by the callers. The metadata in each row does not need to be written as Avro or UnsafeRow,
  * but the actual data provided by the caller does.
  */
+/** Interface for encoding and decoding state store data between UnsafeRow and raw bytes.
+ *
+ * @note All encode methods expect non-null input rows. Handling of null values is left to the
+ * implementing classes.
+ */
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:67]: is this in the right place?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala [hunk 5]
```diff
@@ -60,15 +60,92 @@ sealed trait RocksDBValueStateEncoder {
...
 trait DataEncoder {
+  /** Encodes a complete key row into bytes. Used as the primary key for state lookups.
+   *
+   * @param row An UnsafeRow containing all key columns as defined in the keySchema
+   * @return Serialized byte array representation of the key
+   */
   def encodeKey(row: UnsafeRow): Array[Byte]
+
+  /** Encodes the non-prefix portion of a key row. Used with prefix scan and
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:76]: Text should start at the second line, not the first in scaladoc

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala [hunk 6]
```diff
@@ -402,6 +479,80 @@ class AvroStateEncoder(
...
+   * For prefix scan cases, it also creates separate encoders for the suffix portion of keys.
+   *
+   * @param keyStateEncoderSpec Specification for how to encode keys
+   * @param valueSchema Schema for the values to be encoded
+   * @return An AvroEncoder containing all necessary serializers and deserializers
+   */
+  private def createAvroEnc(
+                             keyStateEncoderSpec: KeyStateEncoderSpec,
+                             valueSchema: StructType
+                           ): AvroEncoder = {
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:515]: please fix the indent

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala [hunk 7]
```diff
@@ -402,6 +479,80 @@ class AvroStateEncoder(
 
   private lazy val remainingKeyAvroProjection = UnsafeProjection.create(remainingKeySchema)
 
+
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:482]: uber nit: delete extra empty lines

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala [hunk 8]
```diff
@@ -789,44 +940,58 @@ abstract class RocksDBKeyStateEncoderBase(
   }
 }
 
+/** Factory object for creating state encoders used by RocksDB state store.
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:943]: move text to next line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 9]
```diff
@@ -325,6 +325,18 @@ sealed trait KeyStateEncoderSpec {
   def keySchema: StructType
   def jsonValue: JValue
   def json: String = compact(render(jsonValue))
+
+  /** Creates a RocksDBKeyStateEncoder for this specification.
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:329]: ditto on scaladoc

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 10]
```diff
@@ -356,6 +376,14 @@ case class PrefixKeyScanStateEncoderSpec(
   if (numColsPrefixKey == 0 || numColsPrefixKey >= keySchema.length) {
     throw StateStoreErrors.incorrectNumOrderingColsForPrefixScan(numColsPrefixKey.toString)
   }
+  override def toEncoder(
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:379]: nit: new line please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceTransformWithStateSuite.scala [hunk 11]
```diff
@@ -126,8 +126,7 @@ class SessionGroupsStatefulProcessorWithTTL extends
  * Test suite to verify integration of state data source reader with the transformWithState operator
  */
 class StateDataSourceTransformWithStateSuite extends StateStoreMetricsTest
-  with AlsoTestWithEncodingTypes
-  with AlsoTestWithChangelogCheckpointingEnabled {
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceTransformWithStateSuite.scala]: did you lose the `AlsoTestWithChangelogCheckpointingEnabled`? Does `AlsoTestWithRocksDBFeatures` contain that?

**Review by brkyvz — CHANGES_REQUESTED** [review]
Basically comment on scaladoc style, otherwise LGTM!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceTransformWithStateSuite.scala (hunk 11 above)
Comment (ericm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceTransformWithStateSuite.scala]: Yeah, AlsoTestWithRocksDBFeatures does test that - not sure why this is showing up in the diff.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala [hunk 12]
```diff
@@ -219,469 +222,402 @@ object RocksDBStateEncoder extends Logging {
...
-              writer.write(idx, bbuf.getDouble)
-            }
-        }
-      }
+  override def encodeRemainingKey(row: UnsafeRow): Array[Byte] = {
+    keyStateEncoderSpec match {
+      case PrefixKeyScanStateEncoderSpec(_, _) =>
+        encodeUnsafeRowToAvro(row, avroEncoder.suffixKeySerializer.get, remainingKeyAvroType, out)
+      case RangeKeyScanStateEncoderSpec(_, _) =>
+        encodeUnsafeRowToAvro(row, avroEncoder.keySerializer, remainingKeyAvroType, out)
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:618]: does this need to be suffixKeySerializer?

Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:618]: No, this is right - RangeKeyScan encoder doesn't use Avro serialization for both parts of the ke

**Review by brkyvz — APPROVED** [review]
LGTM, but you missed one comment: https://github.com/apache/spark/pull/48944/files#r1887807431 on scaladoc. Once done @HeartSaVioR , can you please help merge this?

**Comment by HeartSaVioR** [comment]
The comment link does not show the comment, which I assume the comment is outdated via new commit.
I'm merging on behalf of @brkyvz . Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 13]
```diff
@@ -778,6 +818,7 @@ object StateStore extends Logging {
       storeConf: StateStoreConf,
       hadoopConf: Configuration,
       useMultipleValuesPerKey: Boolean = false): StateStore = {
+    hadoopConf.set(StreamExecution.RUN_ID_KEY, storeProviderId.queryRunId.toString)
```
Comment (Kimahriman) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:821]: Random late comment I noticed in the code, this is already called 4 line below

**Pull request closed without merging**